### PR TITLE
added setExtraParameters to crashReporter

### PR DESCRIPTION
**Pull request by TimvdEijnden**
We want to send along the logged in user in our application when submitting a crash report.
But we want to start the crash report directly on start so we wont miss any crashes.

When we load our website we can get the logged in user from the loaded website and add that info to the extra_parameters of the crashReporter. This was not possible so i've added it.

screen shot 2016-09-02 at 17 08 35

### CODE REVIEW

File: docs/api/crash-reporter.md [hunk 1]
```diff
@@ -58,6 +58,17 @@ sent or the crash reporter has not been started, `null` is returned.
 Returns all uploaded crash reports. Each report contains the date and uploaded
 ID.
 
+### `crashReporter.setExtraParameters(extra)`
+
+* `extra` Object - An object you can define that will be sent along with the
+  report. Only string properties are sent correctly, Nested objects are not
```
Comment (zeke) [inline docs/api/crash-reporter.md]: `correctly, Nested` -> `correctly. Nested`

File: docs/api/crash-reporter.md [hunk 2]
```diff
@@ -58,6 +58,17 @@ sent or the crash reporter has not been started, `null` is returned.
...
+  supported.
+
+Updates the extra value that will be sent along with the report. This allows
```
Comment (zeke) [inline docs/api/crash-reporter.md]: This paragraph is a bit redundant, and could be slimmed down.

**Comment by zeke** [comment]
Nice, @TimvdEijnden. Are you able to add some tests for this new feature?

**Comment by zeke** [comment]
Seeing a failure on Windows CI. It seems unrelated though. Maybe a flaky build?

```
[1228/1228] LINK_EMBED electron.exe
Running test.py
Attempted to load extension "foo" that has already been loaded.
Reading C:\projects\electron\spec\does-not-exist\manifest.json failed.
Error: ENOENT: no such file or directory, open 'C:\projects\electron\spec\does-not-exist\manifest.json'
    at Error (native)
    at Object.fs.openSync (fs.js:640:18)
    at Object.module.(anonymous function) [as openSync] (ELECTRON_ASAR.js:167:20)
    at Object.fs.readFileSync (fs.js:508:33)
    at Object.fs.readFileSync (ELECTRON_ASAR.js:500:29)
    at getManifestFromPath (C:\projects\electron\out\D\resources\electron.asar\browser\chrome-extension.js:33:26)
    at Function.BrowserWindow.addDevToolsExtension (C:\projects\electron\out\D\resources\electron.asar\browser\chrome-extension.js:363:22)
    at callFunction (C:\projects\electron\out\D\resources\electron.asar\browser\rpc-server.js:217:18)
    at EventEmitter.<anonymous> (C:\projects\electron\out\D\resources\electron.asar\browser\rpc-server.js:306:5)
    at emitMany (events.js:127:13)
Parsing C:\projects\electron\spec\fixtures\devtools-extensions\bad-manifest\manifest.json failed.
SyntaxError: Unexpected token } in JSON at position 0
    at Object.parse (native)
    at getManifestFromPath (C:\projects\electron\out\D\resources\electron.asar\browser\chrome-extension.js:41:21)
    at Function.BrowserWindow.addDevToolsExtension (C:\projects\electron\out\D\resources\electron.asar\browser\chrome-extension.js:363:22)
    at callFunction (C:\projects\electron\out\D\resources\electron.asar\browser\rpc-server.js:217:18)
    at EventEmitter.<anonymous> (C:\projects\electron\out\D\resources\electron.asar\browser\rpc-server.js:306:5)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (C:\projects\electron\out\D\resources\electron.asar\browser\api\web-contents.js:218:13)
    at emitTwo (events.js:106:13)
```

File: docs/api/crash-reporter.md (hunk 1 above)
Comment (TimvdEijnden) [inline docs/api/crash-reporter.md]: I've copied it from the start method docs 😉

**Comment by TimvdEijnden** [comment]
I'll try to add tests,  I have zero knowledge of C++ 😇

**Comment by TimvdEijnden** [comment]
I was writing tests to verify if my changes worked, but it doesn't work :( 

The implementation is different per OS, I think it's possible to add this support but I don't have the experience to fix this. It's too difficult for a me.

File: docs/api/crash-reporter.md [hunk 3]
```diff
@@ -58,6 +58,17 @@ sent or the crash reporter has not been started, `null` is returned.
 Returns all uploaded crash reports. Each report contains the date and uploaded
 ID.
 
+### `crashReporter.setExtraParameters(extra)`
```
Comment (kevinsawicki) [inline docs/api/crash-reporter.md]: So these are meant to completely replace the `extra` parameters specified to `start` right? Maybe the docs should mention something like `This will override any extra parameters specified to start()` just to make things clear.

File: spec/api-crash-reporter-spec.js [hunk 4]
```diff
@@ -50,6 +50,7 @@ describe('crashReporter module', function () {
         assert.equal(fields['platform'], process.platform)
         assert.equal(fields['extra1'], 'extra1')
         assert.equal(fields['extra2'], 'extra2')
+        assert.equal(fields['extra3'], 'extra3')
```
Comment (kevinsawicki) [inline spec/api-crash-reporter-spec.js]: Instead of adding this assertion here, it would seem more comprehensive to add a second spec that calls `setExtraParameters`  explicitly with different values for `extra1` and `extra2` to verify that `setExtraParameters overrides the `start` extra params.

Then we'll be sure the extra parameters work in both cases, specified to `start` and set explicitly via `setExtraParameters`.

**Comment by bridiver** [comment]
I've been working on pulling in support `base::debug::SetCrashKeyValue` and will have a PR ready soon. Seems like we could combine the two and add an electron api for `setCrashKeyValue` that would add the values to the crash metadata just like they do in chrome

**Comment by TimvdEijnden** [comment]
Yes that would be great. Adding the setExtraParameters was easy but getting it to work is to hard. 
As long as I can add extra data after starting the crash reporter I'm happy. 

I'm closing this for now and will wait for your PR.

**Comment by zeke** [comment]
> I've been working on pulling in support base::debug::SetCrashKeyValue and will have a PR ready soon. 

Hey @bridiver have you made any progress on this? If not, I can pick it up.

**Comment by bridiver** [comment]
@zeke I have a partially complete branch. I think I should be able to pick it back up and finish, but if not I'll link you to what I have and you can pick it up from there
